### ISSUE TITLE
Make it possible to write Links IR programs

### ISSUE THREAD

**Issue by frank-emrich** [links-lang/links#878]
In the Links meeting we discussed that it may be helpful for different purposes to write Links IR programs directly. One use case would be testing the IR type-checker.

We discussed different ways of implementing this, like the following:
1. Use the existing derived parsers for JSON or s-expression representation. This is easy to do, but yields syntax that is probably very painful to use.
2. We could define a proper syntax for the IR and write/generate a parser for it. This may or may not require implementing a similar parser for a simplified language of Links types. I specifically mention the types here because I would like to be able to write types in a way so that we have more direct control about their internal representation (e.g., being able to directly specify the integer identifiers for type variables).
3. We could use a mixture of the two.

**Comment by jstolarek** [comment]
While I like the idea of being to supply an externally-generated IR to the IR compiler (this creates a possibility for writing a new frontend that works with Links' IR backend), I am *very* sceptical about writing IR programs directly (in a sense of having a language that is parsed to the IR). What is the exact motivation/use case for having such a feature? If `links` library interface exposed the IR AST and the entry point to the IR compiler/typechecker, we could write IR programs by building their ASTs directly, with no need for any parsing. This would allow to test the IR typechecker fairly easily.

There is a similar feature in GHC but the idea of writing Core programs by hand never took off. See [An External Representation for the GHC Core Language (For GHC 6.10)](http://www.haskell.org/ghc/docs/6.10.4/html/ext-core/core.pdf) paper and [GHC wiki](https://gitlab.haskell.org/ghc/ghc/-/wikis/external-core).

**Comment by rudihorn** [comment]
When used directly ocaml can also be pretty good at writing small embedded DSLs for defining these types of things. I've used the `let` syntax for defining mini lambda calculi which has worked quite well.

e.g. the following example I used to compile a lambda abstraction `\x -> (let y = x in if true then 20 else y)`. It's not perfect, but for internal purposes it can work well.
```ocaml
  let open Lang.O in
  abs1 "x" (fun x ->
      let@ y = x in
      ifelse ctrue (i 20) y)
```

Regarding james' PPX question: Because ocaml just derives the `pp` function it is easy to override behaviour to a certain extent, but becomes problematic in the recursive case. The following example might make what I mean more obvious:

```ocaml
(* derived function *)
let rec pp x = ... pp subterm ...

(* our function *)
let pp_ours x = special case of x | pp x
```

**Comment by frank-emrich** [comment]
@jamescheney 
>  Would allowing writing "raw" integer variables in the IR concrete syntax suffice (i.e. something morally equivalent to the current derived IR pretty-printing?) Or do you have something else in mind? 

Yes, that's what I had in mind. I mostly just want to be able to have as much control about the resulting representation of the types as possible.

Regarding your point (2):
There is no general answer to this, but most deriving plugins allow manually overriding the behavior; in particular the ones we use (e.g., show, yojson). So it would be possible to wire together a "parser" for IR programs that reads types and parses them using our existing parser for types, and parses the rest of the IR programs using a generated parser accepting JSON syntax directly corresponding to the underlying IR types. I'm not saying this is the best solution to go for, but this would be technically possible.


@jstolarek 

My motivation for doing is that I would like to be able to test the IR type-checker more independently from the frontend. There have been many times when I fixed something in it, but it was either complicated or impossible to write a Links program whose translation would look exactly like what I wanted. This holds in particular for small regression tests a la "the IR type-checker previously messed up something about type variable scoping here, let's write a test case or two to make sure it doesn't break again" because at the moment there is a good chance that the Links program you write will fail to type-check for unrelated reasons.

I agree with your point (and James' summary in (1)) that writing such programs is very painful, possibly *too* painful to justify the effort required to make it possible, unless we have a decent parser for it that let's you write it in a more readable form. Following the feedback from @rudihorn, I will give a DSL-based approach a try. This won't be too different from writing the painful JSON representation, but at least it would be statically type-checked and potentially more easy to refactor.

In summary, I suggest that my next step will be to create a small DSL for writing Links IR programs, I will use it to write a few test cases and then we can discuss whether this gets the job done or we want to do something more sophisticated (like an actual parser). If we were to go down this path, the IR tests would be more similar to unit tests and I would explore how to integrate them with the existing test suite.

**Comment by jstolarek** [comment]
@frank-emrich, I believe we have the same thing in mind. I was thinking about exposing the ability to construct IR ASTs, which is exactly what a DSL approach will achieve (possibly hiding some of the plumbing).
